### PR TITLE
`MetaBackend`

### LINKED ISSUE #1096: List available backends
It would be nice if Qibo were able to check if a backend is available, and have a function to survey all the registered backends, and list those available.

Possibly, it could be implemented even trying to load it, and catching the failure. But if each backend had a property `available` it would be even better.
(since the backend might know about the required conditions for availability, and we'd avoid catching a generic error)

### PR DESCRIPTION
**Pull request by BrunoLiegiBastonLiegi**
This PR implements a `MetaBackend` object that will call the different projects' (`qibojit`, `qibolab`, etc...) `MetaBackend` to load their backends. This will replace the current `construct_backend` function and will also give access to a `list_available()` function to list all the backends available in the system.

Closes #1096.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

File: src/qibo/backends/__init__.py
Comment (alecandido) [inline src/qibo/backends/__init__.py:42]: I'm not sure whether I'd nest that here.

Instead, we could make a separate function, and the Qibo `MetaBackend` could be just one like anyone else, providing a nice example of how to implement one.

Eventually, we could discover backends with some runtime mechanism, without relying on something like
https://github.com/qiboteam/qibo/blob/f43a6a1a7eea95c0da1ecb65a9c6d346ecfe0a0a/src/qibo/backends/__init__.py#L13
(e.g. by having installed additional packages prefixed with `qibo-backend-` or something like that).
But I'd keep it simple for this PR, just properly lay the ground.

File: src/qibo/backends/__init__.py
Comment (alecandido) [inline src/qibo/backends/__init__.py:65]: Same as above.

File: src/qibo/backends/__init__.py
Comment (alecandido) [inline src/qibo/backends/__init__.py:206]: These are the functions that could implement the mechanism above.

File: tests/test_models_circuit_backpropagation.py
Comment (alecandido) [inline tests/test_models_circuit_backpropagation.py:13]: Here, and in the other tests, I'd keep the `construct_backend` function.

The reason is partly the one above, but even that we can simply avoid to break, and keep using the old function. Since, in this case, it costs nothing to us, I'd just do that.

**Review by alecandido — COMMENTED** [review]
A suggestion about how to rearrange the non-native backends. But other than that, it looks perfect.

Thanks @BrunoLiegiBastonLiegi to have taken care to implement this idea everywhere :D

**Comment by BrunoLiegiBastonLiegi** [comment]
We have the same problem here as in #1239. Namely, since now in main `qibo` depends on `qibotn/py3.12` and `qibojit/py3.12`, I cannot resolve the conflicts without breaking something. For `qibojit` there is a simple solution which is changing the `qibo` dependency to `qibojit/main` as the corresponding `qibojit` PR has already been merged, however for `qibotn` the `py3.12` PR does not seem to be mergeable soon.

**Comment by alecandido** [comment]
True, this py3.12 upgrade is being a bit of a nightmare (I hope `qibo-core` will land soon...)

> however for `qibotn` the `py3.12` PR does not seem to be mergeable soon.

Here I disagree a bit, since https://github.com/qiboteam/qibotn/pull/48 should be as good as `main`, and py3.12 compatible.
I just need the two approvals, thus I'm adding you as well as a reviewer :P

**Review by stavros11 — APPROVED** [review]

File: pyproject.toml
Comment (stavros11) [inline pyproject.toml:72]: I think https://github.com/qiboteam/qibojit/pull/172 can be merged. So maybe we should merge the PR in qibojit first and point this back to main before merging?

File: poetry.lock
Comment (alecandido) [inline poetry.lock:3561]: I fear the problem is coming from here, but not sure. I'd try to pin Pylint to `3.1.0` in `pyproject.toml` and regenerate the lock.

If this is the actual reason, we could merge with the Pylint version pinned, but we should also consider fixing the errors.

Comment (alecandido) [reply poetry.lock]: I applied the suggestion myself in a83e0c2, let's what happens

Comment (alecandido) [reply poetry.lock]: Ok, hypothesis confirmed:
https://github.com/qiboteam/qibo/actions/runs/9099789800/job/25013165174#step:10:16
```py
src/qibo/backends/__init__.py:21:40: E0602: Undefined variable 'Backend' (undefined-variable)
src/qibo/backends/__init__.py:205:44: E0602: Undefined variable 'Backend' (undefined-variable)
```

Now the error is sensible, since related to this PR, and it should be fixed (it's just a missing import).

However, it could be worth to fix the previous ones as well.

**Comment by alecandido** [comment]
@BrunoLiegiBastonLiegi the issue with Pylint 3.2 is not completely trivial, thus I opened #1328 for it.

At this point, it's not worth to stop this any longer. Feel free to merge whenever you wish :)

File: poetry.lock
Comment (alecandido) [reply poetry.lock]: #1328

**Comment by BrunoLiegiBastonLiegi** [comment]
I was waiting for `qibotn` and `qibo-cloud-backends` to be merged before merging this one, just to avoid breaking the `set_backend` for those two.

**Pull request merged**
